### DISCUSSION TITLE
Writing to EEPROM oddly

### DISCUSSION THREAD

**Discussion by sirquackiii** [TomNisbet/TommyPROM#35] [Q&A] [answered]
I made the TommyPROM design on a breadboard, and it seemed to work.
When I uploaded a file, no errors popped up.
So when I wrote my simple assembly program to the chip, it only wrote EAs (nop instructions).

I thought this was because it was write protected, so I unlocked it and uploaded again, but same problem.
This is what TommyPROM dumps:
![image](https://user-images.githubusercontent.com/77940766/141203806-9d495c11-c51f-4b46-a9a1-7170100a3ba2.png)

This the the binary file I uploaded:
![image](https://user-images.githubusercontent.com/77940766/141203860-90453c9c-f5a7-4165-82fe-2622c2809f88.png)

But the oddest thing is when I hook up the EEPROM to the 6502 project itself, it reads seemingly random bytes before getting to the NOPs (EA).

When I ran the zap, it gave out an error, saying that it read `0x33` instead of `0x41`.

I decided to check out the binary of the two of the numbers to see if 2 bits were flipped or something, but there was no sign of any wiring issues:
```
0x33: 00110011
0x41: 01000001
```

What's the problem here?

Thanks.

**Comment by TomNisbet** [comment, marked as answer]
It's hard to debug because you can't really trust that the data from the **D** command is accurate either.  I suspect that the code is writing to the chip, but the addressing is wrong.  33 and 41 are both part of the zap pattern and EA is in the values that you write.  Check that you don't have any address lines swapped or shorted and also check the clock and data lines to the shift registers.

The [troubleshooting guide](https://tomnisbet.github.io/TommyPROM/troubleshooting/) has some tips for this, particularly the **A** command.  You should be able to write some different address values and verify that they are being presented on the correct pins at the EEPROM.

**Reply by sirquackiii** [reply]
I think I found the problem.
I set the address wires to FFFF, and I tested them with the multimeter. All of them were low! I think it was a problem with the shift registers, I wired something incorrectly which stopped the Arduino from communicating with the address lines. This explains everything being EA, since the last byte in the binary file is EA, and the Arduino could only read address 0, hence everything being EA. Currently rewiring.

**Reply by sirquackiii** [reply]
That helped, but now I have a different problem. Now, it is successfully writing EAs, except not the starting code, and some sections are FFs for some reason, and multiple bytes are seemingly random. I found one possible problem, which is the A6 pin. A6 never goes high even when I set all the address lines to high in the debugging tool. I checked for physical damage to the pin, and there was none. I checked the connections between the shift register and the pin but it all seemed fine. Could one pin not work despite everything else working?

**Reply by TomNisbet** [reply]
I can't think of anything special about the A6 pin.  Try pulling the EEPROM and checking the address lines using the debug tool.  If you write addresses 80, you should see all-zero on the address lines except for a one at A7.  Then try address 40 and 20 and you should see the same thing with a single one at A6 and A5.  If A7 and A5 work normally but A6 stays low then I'd suspect that the problem is with that particular pin, so check the wiring and make sure it isn't shorted.  You can also swap the two shift registers and use addresses 8000, 4000, and 2000 to test A15 thru A13. 
Also try testing A6 right at the shift register instead of at the EEPROM.  You can pull the wire to the EEPROM to make sure there isn't a short.
